src/osgUtil/MeshOptimizers.cpp

```diff
@@ -98,40 +100,62 @@ struct GeometryArrayGatherer
     }
 
     ArrayList _arrayList;
+    osg::ref_ptr<osg::Vec3Array> _vertexarray;
 };
 
 // Compare vertices in a mesh using all their attributes. The vertices
-// are identified by their index. Extracted from TriStripVisitor.cpp
+// are identified by their index.
 struct VertexAttribComparitor : public GeometryArrayGatherer
 {
+    float _epsilondist;
     VertexAttribComparitor(osg::Geometry& geometry)
-        : GeometryArrayGatherer(geometry)
+        : GeometryArrayGatherer(geometry), _epsilondist(1e-9)
     {
     }
 
+    ///sort by distance to origin 
     bool operator() (unsigned int lhs, unsigned int rhs) const
     {
-        for(ArrayList::const_iterator itr=_arrayList.begin();
-            itr!=_arrayList.end();
+        for(ArrayList::const_reverse_iterator itr=_arrayList.rbegin();
+            itr!=_arrayList.rend();
             ++itr)
         {
             int compare = (*itr)->compare(lhs,rhs);
             if (compare==-1) return true;
             if (compare==1) return false;
         }
+        if (_vertexarray.valid())
+        {
+            float dpos1 = _vertexarray->at(lhs).length2();
+            float dpos2 = _vertexarray->at(rhs).length2();
```

Comment:
Computing the length2() of each vertex for each comparison is expensive.  The efficient way if you really wanted to do something like this would be compute the control distance in one pass and store in a single cache vector.